### ISSUE TITLE
[clang-format] ColumnLimit is not respected in requires clause with BreakBeforeBinaryOperators: NonAssignment

### ISSUE THREAD

**Issue by mkurdej** [llvm/llvm-project#54057]
Tested with a fresh clang-format from main branch at commit bfb4afee74c8d6e3b1d020564bfe163073f07a04 (2022-02-24):
```
clang-format version 15.0.0 (git@github.com:llvm/llvm-project.git bfb4afee74c8d6e3b1d020564bfe163073f07a04)
```

Given config:
```
BasedOnStyle: LLVM
BreakBeforeBinaryOperators: NonAssignment
```
the code is formatted to:
```
template <typename T>
  requires is_default_constructible_v<
               hash<T>> and is_copy_constructible_v<hash<T>> and is_move_constructible_v<hash<T>> and is_copy_assignable_v<hash<T>> and is_move_assignable_v<hash<T>> and is_destructible_v<hash<T>> and is_swappable_v<hash<T>> and is_callable_v<hash<T>(T)> and is_same_v<size_t, decltype(hash<T>(declval<T>()))> and is_same_v<size_t, decltype(hash<T>(declval<T &>()))> and is_same_v<size_t, decltype(hash<T>(declval<const T &>()))>
struct S {};
```
which is way above the `ColumnLimit: 80` (default for LLVM style).

It seems that the output changes a bit when `IndentWidth` is changed, but it still goes above the `ColumnLimit`:
```
BasedOnStyle: LLVM
BreakBeforeBinaryOperators: NonAssignment
IndentWidth: 4
```
results in:
```
template <typename T>
    requires is_default_constructible_v<hash<T>> and is_copy_constructible_v<hash<T>> and is_move_constructible_v<hash<T>> and is_copy_assignable_v<hash<T>> and is_move_assignable_v<hash<T>> and is_destructible_v<hash<T>> and is_swappable_v<hash<T>> and is_callable_v<hash<T>(T)> and is_same_v<size_t, decltype(hash<T>(declval<T>()))> and is_same_v<size_t, decltype(hash<T>(declval<T &>()))> and is_same_v<size_t, decltype(hash<T>(declval<const T &>()))>
struct S {};
```

There are other strange results:
```
BasedOnStyle: LLVM
BreakBeforeBinaryOperators: NonAssignment
ColumnLimit: 40
```
gives a requires clause on a single line too:
```
template <typename T>
  requires is_default_constructible_v<hash<T>> and is_copy_constructible_v<hash<T>> and is_move_constructible_v<hash<T>> and is_copy_assignable_v<hash<T>> and is_move_assignable_v<hash<T>> and is_destructible_v<hash<T>> and is_swappable_v<hash<T>> and is_callable_v<hash<T>(T)> and is_same_v<size_t, decltype(hash<T>(declval<T>()))> and is_same_v<size_t, decltype(hash<T>(declval<T &>()))> and is_same_v<size_t, decltype(hash<T>(declval<const T &>()))>
struct S {};
```

FYI, @HazardyKnusperkeks 

Expected code should respect the `ColumnLimit` of course.

**Comment by mkurdej** [comment]
It seems that the results are somewhat similar with `BreakBeforeBinaryOperators: All`.
Only `BreakBeforeBinaryOperators: None` behaves correctly.

**Comment by HazardyKnusperkeks** [comment]
I'll take a look at it.

**Comment by HazardyKnusperkeks** [comment]
This is not limited to requires clauses, until now just no one had the options with a template closer on the lhs. I'll fix it nevertheless.

**Comment by HazardyKnusperkeks** [comment]
Review: https://reviews.llvm.org/D120621

**Comment by HazardyKnusperkeks** [comment]
Fixed in d6daca21738aed64837b825c32d4f1b4f54cec04, which has the wrong (but also fixed) issue in the message.
